cluster/kubernetes/sync.go

```diff
@@ -2,36 +2,327 @@ package kubernetes
 
 import (
 	"bytes"
+	"crypto/sha1"
+	"encoding/hex"
 	"fmt"
 	"io"
 	"os/exec"
 	"sort"
 	"strings"
 	"time"
 
-	rest "k8s.io/client-go/rest"
-
 	"github.com/go-kit/kit/log"
+	"github.com/imdario/mergo"
 	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	rest "k8s.io/client-go/rest"
+
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/cluster"
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+const (
+	stackLabel         = kresource.PolicyPrefix + "stack"
+	checksumAnnotation = kresource.PolicyPrefix + "stack_checksum"
+
+	DefaultDefaultNamespace = "default"
 )
 
+// Sync takes a definition of what should be running in the cluster,
+// and attempts to make the cluster conform. An error return does not
+// necessarily indicate complete failure; some resources may succeed
+// in being synced, and some may fail (for example, they may be
+// malformed).
+func (c *Cluster) Sync(spec cluster.SyncDef) error {
+	logger := log.With(c.logger, "method", "Sync")
+
+	fallbackNamespace := c.applier.getDefaultNamespace()
+
+	type checksum struct {
+		stack, sum string
+	}
+	// Keep track of the checksum of each stack, so we can compare
+	// them during garbage collection.
+	checksums := map[string]checksum{}
+
+	// NB we get all resources, since we care about leaving unsynced,
+	// _ignored_ resources alone.
+	clusterResources, err := c.getResourcesBySelector("", fallbackNamespace)
+	if err != nil {
+		return errors.Wrap(err, "collating resources in cluster for sync")
+	}
+
+	cs := makeChangeSet()
+	var errs cluster.SyncError
+	for _, stack := range spec.Stacks {
+		for _, res := range stack.Resources {
+			id := res.ResourceID().String()
+			// make a record of the checksum, whether we stage it to
+			// be applied or not, so that we don't delete it later.
+			csum := sha1.Sum(res.Bytes())
+			checkHex := hex.EncodeToString(csum[:])
+			checksums[id] = checksum{stack.Name, checkHex}
+			if res.Policy().Has(policy.Ignore) {
+				logger.Log("info", "not applying resource; ignore annotation in file", "resource", res.ResourceID(), "source", res.Source())
+				continue
+			}
+			if cres, ok := clusterResources[id]; ok && cres.Policy().Has(policy.Ignore) {
+				logger.Log("info", "not applying resource; ignore annotation in cluster resource", "resource", cres.ResourceID())
+				continue
+			}
+			resBytes, err := applyMetadata(res, stack.Name, checkHex)
+			if err == nil {
+				cs.stage("apply", res.ResourceID(), res.Source(), resBytes)
```

Comment:
Unrelated to this PR, but ... is there a good reason to use strings directly instead of constants?

Reply:
.. or even different methods -- I think there's a bit of faux generality in here.